mac build failed

Use 'make build_linux'  on mac, occued error.
error meassage :
docker run --rm \
		-v :/go \
		-w /go/src/github.com/chaosblade-io/chaosblade \
		chaosblade-build-musl:latest
make: *** No targets specified and no makefile found.  Stop.
make: *** [build_linux] Error 2

I have to install go env?

> Use 'make build_linux' on mac, occued error.
> error meassage :
> docker run --rm 
> -v :/go 
> -w /go/src/github.com/chaosblade-io/chaosblade 
> chaosblade-build-musl:latest
> make: *** No targets specified and no makefile found. Stop.
> make: *** [build_linux] Error 2
> 
> I have to install go env?

Not necessary. Did you compile the project or the release package?
The release package does not need to compile.

First, I clone the code, make build_linux  on mac, occured error. 
error message:error meassage :
docker run --rm
-v :/go
-w /go/src/github.com/chaosblade-io/chaosblade
chaosblade-build-musl:latest
make: *** No targets specified and no makefile found. Stop.
make: *** [build_linux] Error 2

Then ,I download  0.0.1 release package,  run ./blade , " cannot execute binary file"

env:
MACOS MOJAVE 10.14.2
chaosblade v0.0.1

> First, I clone the code, make build_linux on mac, occured error.
> error message:error meassage :
> docker run --rm
> -v :/go
> -w /go/src/github.com/chaosblade-io/chaosblade
> chaosblade-build-musl:latest
> make: *** No targets specified and no makefile found. Stop.
> make: *** [build_linux] Error 2
> 
> Then ,I download 0.0.1 release package, run ./blade , " cannot execute binary file"
> 
> env:
> MACOS MOJAVE 10.14.2
> chaosblade v0.0.1

Just try my PR for build on MacOS, type `make docker_image` to build.

--- a/exec/local.go
+++ b/exec/local.go
@@ -45,7 +45,8 @@ func execScript(ctx context.Context, script, args string) *transport.Response {
 
 // GetPidsByProcessCmdName
 func GetPidsByProcessCmdName(processName string, ctx context.Context) ([]string, error) {
-	response := channel.Run(ctx, "pgrep", fmt.Sprintf(`-l %s | grep -v -w blade | grep -v -w chaos_killprocess | awk '{print $1}' | tr '\n' ' '`, processName))
+	response := channel.Run(ctx, "pgrep",
+		fmt.Sprintf(`-l %s | grep -v -w blade | grep -v -w chaos_killprocess | awk '{print $1}' | tr '\n' ' '`, processName))
 	if !response.Success {
 		return nil, fmt.Errorf(response.Err)
 	}
@@ -64,11 +65,12 @@ func GetPidsByProcessName(processName string, ctx context.Context) ([]string, er
 	if otherProcess != nil {
 		processString := otherProcess.(string)
 		if processString != "" {
-			otherGrepInfo = fmt.Sprintf("| grep %s", processString)
+			otherGrepInfo = fmt.Sprintf(`| grep "%s"`, processString)
 		}
 	}
-	response := channel.Run(ctx, "ps", fmt.Sprintf(`%s | grep %s %s | grep -v -w grep | grep -v -w blade | grep -v -w chaos_killprocess | awk '{print $2}' | tr '\n' ' '`,
-		psArgs, processName, otherGrepInfo))
+	response := channel.Run(ctx, "ps",
+		fmt.Sprintf(`%s | grep %s %s | grep -v -w grep | grep -v -w blade | grep -v -w chaos_killprocess | awk '{print $2}' | tr '\n' ' '`,
+			psArgs, processName, otherGrepInfo))
 	if !response.Success {
 		return nil, fmt.Errorf(response.Err)
 	}
@@ -90,4 +92,10 @@ func GetPsArgs(ctx context.Context) string {
 		psArgs = "-o user,pid,ppid,args"
 	}
 	return psArgs
+}
+
+// IsCommandAvailable return true if the command exists
+func IsCommandAvailable(commandName string) bool {
+	response := execScript(context.TODO(), "command", fmt.Sprintf("-v %s", commandName))
+	return response.Success
 }

--- a/exec/os/bin/burncpu.go
+++ b/exec/os/bin/burncpu.go
@@ -4,28 +4,37 @@ import (
 	"context"
 	"flag"
 	"fmt"
-	"github.com/chaosblade-io/chaosblade/exec"
-	"github.com/chaosblade-io/chaosblade/util"
-	"path"
 	"runtime"
 	"strings"
 	"time"
+	"path"
+
+	"github.com/chaosblade-io/chaosblade/util"
+	"github.com/chaosblade-io/chaosblade/exec/os"
+	"github.com/chaosblade-io/chaosblade/exec"
+	"strconv"
 )
 
 var (
 	burnCpuStart, burnCpuStop, burnCpuNohup bool
-	numCPU                                  int
+	cpuCount                                int
+	cpuList                                 string
+	cpuProcessor                            string
 )
 
+const cpuProcessorFlag = "cpu-processor"
+
 func main() {
-	flag.BoolVar(&burnCpuStart, "start", false, "burn cpu")
-	flag.BoolVar(&burnCpuStop, "stop", false, "stop burn cpu")
-	flag.BoolVar(&burnCpuNohup, "nohup", false, "nohup to run burn cpu")
-	flag.IntVar(&numCPU, "numcpu", runtime.NumCPU(), "number of cpus")
+	flag.BoolVar(&burnCpuStart, os.StartFlag, false, "burn cpu")
+	flag.BoolVar(&burnCpuStop, os.StopFlag, false, "stop burn cpu")
+	flag.StringVar(&cpuList, os.CpuListFlag, "", "CPUs in which to allow burning (1,3)")
+	flag.BoolVar(&burnCpuNohup, os.NohupFlag, false, "nohup to run burn cpu")
+	flag.IntVar(&cpuCount, os.CpuCountFlag, runtime.NumCPU(), "number of cpus")
+	flag.StringVar(&cpuProcessor, cpuProcessorFlag, "0", "only used for identifying process of cpu burn")
 	flag.Parse()
 
-	if numCPU <= 0 || numCPU > runtime.NumCPU() {
-		numCPU = runtime.NumCPU()
+	if cpuCount <= 0 || cpuCount > runtime.NumCPU() {
+		cpuCount = runtime.NumCPU()
 	}
 
 	if burnCpuStart {
@@ -38,9 +47,9 @@ func main() {
 }
 
 func burnCpu() {
-	runtime.GOMAXPROCS(numCPU)
+	runtime.GOMAXPROCS(cpuCount)
 
-	for i := 0; i < numCPU; i++ {
+	for i := 0; i < cpuCount; i++ {
 		go func() {
 			for {
 				for i := 0; i < 2147483647; i++ {
@@ -52,33 +61,84 @@ func burnCpu() {
 	select {} // wait forever
 }
 
-const burnCpuBin = "chaos_burncpu"
-
 // startBurnCpu by invoke burnCpuBin with --nohup flag
 func startBurnCpu() {
-	args := fmt.Sprintf(`%s --nohup --numcpu %d > /dev/null 2>&1 &`,
-		path.Join(util.GetProgramPath(), burnCpuBin), numCPU)
 	ctx := context.Background()
-	response := exec.NewLocalChannel().Run(ctx, "nohup", args)
+	if cpuList != "" {
+		cpuCount = 1
+		cores := strings.Split(cpuList, os.CommaDelimiter)
+		for _, core := range cores {
+			pid := runBurnCpu(ctx, cpuCount, true, core)
+			bindBurnCpu(ctx, core, pid)
+		}
+	} else {
+		runBurnCpu(ctx, cpuCount, false, "")
+	}
+	checkBurnCpu(ctx)
+}
+
+// runBurnCpu
+func runBurnCpu(ctx context.Context, cpuCount int, pidNeeded bool, processor string) int {
+	args := fmt.Sprintf(`%s --%s --%s %d`,
+		path.Join(util.GetProgramPath(), os.BurnCpuCommand), os.NohupFlag, os.CpuCountFlag, cpuCount)
+
+	if pidNeeded {
+		args = fmt.Sprintf("%s --%s %s", args, cpuProcessorFlag, processor)
+	}
+	args = fmt.Sprintf(`%s > /dev/null 2>&1 &`, args)
+	response := exec.NewLocalChannel().Run(ctx, os.NohupCommand, args)
 	if !response.Success {
+		stopBurnCpu()
 		printErrAndExit(response.Err)
 	}
+	if pidNeeded {
+		// parse pid
+		newCtx := context.WithValue(context.Background(), exec.ProcessKey, fmt.Sprintf("%s %s", cpuProcessorFlag, processor))
+		pids, err := exec.GetPidsByProcessName(os.BurnCpuCommand, newCtx)
+		if err != nil {
+			stopBurnCpu()
+			printErrAndExit(fmt.Sprintf("bind cpu core failed, cannot get the burning program pid, %v", err))
+		}
+		if len(pids) > 0 {
+			// return the first one
+			pid, err := strconv.Atoi(pids[0])
+			if err != nil {
+				stopBurnCpu()
+				printErrAndExit(fmt.Sprintf("bind cpu core failed, get pid failed, pids: %v, err: %v", pids, err))
+			}
+			return pid
+		}
+	}
+	return -1
+}
+
+// bindBurnCpu by taskset command
+func bindBurnCpu(ctx context.Context, core string, pid int) {
+	response := exec.NewLocalChannel().Run(ctx, os.TasksetCommand, fmt.Sprintf("-cp %s %d", core, pid))
+	if !response.Success {
+		stopBurnCpu()
+		printErrAndExit(response.Err)
+	}
+}
+
+// checkBurnCpu
+func checkBurnCpu(ctx context.Context) {
 	time.Sleep(time.Second)
 	// query process
-	ctx = context.WithValue(ctx, exec.ProcessKey, "nohup")
-	pids, _ := exec.GetPidsByProcessName(burnCpuBin, ctx)
+	ctx = context.WithValue(ctx, exec.ProcessKey, os.NohupCommand)
+	pids, _ := exec.GetPidsByProcessName(os.BurnCpuCommand, ctx)
 	if pids == nil || len(pids) == 0 {
-		printErrAndExit(fmt.Sprintf("%s pid not found", burnCpuBin))
+		printErrAndExit(fmt.Sprintf("%s pid not found", os.BurnCpuCommand))
 	}
 }
 
 // stopBurnCpu
 func stopBurnCpu() {
 	// add grep nohup
-	ctx := context.WithValue(context.Background(), exec.ProcessKey, "nohup")
-	pids, _ := exec.GetPidsByProcessName(burnCpuBin, ctx)
+	ctx := context.WithValue(context.Background(), exec.ProcessKey, os.NohupCommand)
+	pids, _ := exec.GetPidsByProcessName(os.BurnCpuCommand, ctx)
 	if pids == nil || len(pids) == 0 {
-		printOutputAndExit("pid not found")
+		return
 	}
-	exec.NewLocalChannel().Run(ctx, "kill", fmt.Sprintf(`-9 %s`, strings.Join(pids, " ")))
+	exec.NewLocalChannel().Run(ctx, os.KillCommand, fmt.Sprintf(`-9 %s`, strings.Join(pids, " ")))
 }

--- a/exec/os/common.go
+++ b/exec/os/common.go
@@ -1,0 +1,24 @@
+package os
+
+const (
+	StartFlag = "start"
+	StopFlag  = "stop"
+	NohupFlag = "nohup"
+)
+
+// chaosblade command
+const (
+	BurnCpuCommand = "chaos_burncpu"
+)
+
+// system command
+const (
+	NohupCommand   = "nohup"
+	TasksetCommand = "taskset"
+	KillCommand    = "kill"
+)
+
+// delimiter
+const (
+	CommaDelimiter = ","
+)

--- a/exec/os/cpu.go
+++ b/exec/os/cpu.go
@@ -9,6 +9,12 @@ import (
 
 	"github.com/chaosblade-io/chaosblade/exec"
 	"github.com/chaosblade-io/chaosblade/transport"
+	"strings"
+)
+
+const (
+	CpuCountFlag = "cpu-count"
+	CpuListFlag  = "cpu-list"
 )
 
 type CpuCommandModelSpec struct {
@@ -39,8 +45,13 @@ func (*CpuCommandModelSpec) Actions() []exec.ExpActionCommandSpec {
 func (cms *CpuCommandModelSpec) Flags() []exec.ExpFlagSpec {
 	return []exec.ExpFlagSpec{
 		&exec.ExpFlag{
-			Name:     "numcpu",
-			Desc:     "number of cpus",
+			Name:     CpuCountFlag,
+			Desc:     "Cpu count",
+			Required: false,
+		},
+		&exec.ExpFlag{
+			Name:     CpuListFlag,
+			Desc:     "CPUs in which to allow burning (0-3 or 1,3)",
 			Required: false,
 		},
 	}
@@ -103,37 +114,94 @@ func (ce *cpuExecutor) SetChannel(channel exec.Channel) {
 }
 
 func (ce *cpuExecutor) Exec(uid string, ctx context.Context, model *exec.ExpModel) *transport.Response {
-	// number of cpu cores
-	numcpuStr := model.ActionFlags["numcpu"]
-	var numcpu uint64
-	var err error
-	if numcpuStr != "" {
-		numcpu, err = strconv.ParseUint(numcpuStr, 10, 64)
-		if err != nil {
-			return transport.ReturnFail(transport.Code[transport.IllegalParameters], "--numcpu value must be a positive integer")
-		}
-
-	}
-	if numcpu <= 0 || int(numcpu) > runtime.NumCPU() {
-		numcpu = uint64(runtime.NumCPU())
-	}
 	if ce.channel == nil {
 		return transport.ReturnFail(transport.Code[transport.ServerError], "channel is nil")
 	}
 	if _, ok := exec.IsDestroy(ctx); ok {
 		return ce.stop(ctx)
 	} else {
-		return ce.start(ctx, int(numcpu))
+		var cpuCount int
+		var cpuList string
+
+		cpuListStr := model.ActionFlags[CpuListFlag]
+		if cpuListStr != "" {
+			if !exec.IsCommandAvailable(TasksetCommand) {
+				return transport.ReturnFail(transport.Code[transport.EnvironmentError],
+					fmt.Sprintf("%s command not exist", TasksetCommand))
+			}
+			cores, err := parseCpuList(cpuListStr)
+			if err != nil {
+				return transport.ReturnFail(transport.Code[transport.IllegalParameters],
+					fmt.Sprintf("parse %s flag err, %v", CpuListFlag, err))
+			}
+			cpuList = strings.Join(cores, ",")
+		} else {
+			// if cpu-list value is not empty, then the cpu-count flag is invalid
+			var err error
+			cpuCountStr := model.ActionFlags[CpuCountFlag]
+			if cpuCountStr != "" {
+				cpuCount, err = strconv.Atoi(cpuCountStr)
+				if err != nil {
+					return transport.ReturnFail(transport.Code[transport.IllegalParameters],
+						fmt.Sprintf("--%s value must be a positive integer", CpuCountFlag))
+				}
+			}
+			if cpuCount <= 0 || int(cpuCount) > runtime.NumCPU() {
+				cpuCount = runtime.NumCPU()
+			}
+		}
+		return ce.start(ctx, cpuList, cpuCount)
 	}
 }
 
 const burnCpuBin = "chaos_burncpu"
 
-func (ce *cpuExecutor) start(ctx context.Context, numcpu int) *transport.Response {
-	return ce.channel.Run(ctx, path.Join(ce.channel.GetScriptPath(), burnCpuBin),
-		fmt.Sprintf("--start --numcpu %d", numcpu))
+// start burn cpu
+func (ce *cpuExecutor) start(ctx context.Context, cpuList string, cpuCount int) *transport.Response {
+	args := fmt.Sprintf("--%s --%s %d", StartFlag, CpuCountFlag, cpuCount)
+	if cpuList != "" {
+		args = fmt.Sprintf("%s --%s %s", args, CpuListFlag, cpuList)
+	}
+	return ce.channel.Run(ctx, path.Join(ce.channel.GetScriptPath(), burnCpuBin), args)
 }
 
+// stop burn cpu
 func (ce *cpuExecutor) stop(ctx context.Context) *transport.Response {
-	return ce.channel.Run(ctx, path.Join(ce.channel.GetScriptPath(), burnCpuBin), "--stop")
+	return ce.channel.Run(ctx, path.Join(ce.channel.GetScriptPath(), burnCpuBin), fmt.Sprintf("--%s", StopFlag))
+}
+
+// parseCpuList returns the cpu core count. 0,2-3
+func parseCpuList(cpuListValue string) ([]string, error) {
+	cores := make([]string, 0)
+	commaParts := strings.Split(cpuListValue, ",")
+	for _, part := range commaParts {
+		value := strings.TrimSpace(part)
+		if value == "" {
+			continue
+		}
+		if !strings.Contains(value, "-") {
+			_, err := strconv.Atoi(value)
+			if err != nil {
+				return cores, fmt.Errorf("%s value is illegal, %v", value, err)
+			}
+			cores = append(cores, value)
+			continue
+		}
+		coreRange := strings.Split(value, "-")
+		if len(coreRange) != 2 {
+			return cores, fmt.Errorf("%s value is illegal", value)
+		}
+		startIndex, err := strconv.Atoi(strings.TrimSpace(coreRange[0]))
+		if err != nil {
+			return cores, fmt.Errorf("start in %s value is illegal", value)
+		}
+		endIndex, err := strconv.Atoi(strings.TrimSpace(coreRange[1]))
+		if err != nil {
+			return cores, fmt.Errorf("end in %s value is illegal", value)
+		}
+		for i := startIndex; i <= endIndex; i++ {
+			cores = append(cores, strconv.Itoa(i))
+		}
+	}
+	return cores, nil
 }

